clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,445 @@
+        new VeniceProperties(partitionerProperties));
+    return constructVeniceWriter(
+        veniceWriterProperties,
+        new VeniceWriterOptions.Builder(versionCreationResponse.getKafkaTopic()).setPartitioner(venicePartitioner)
+            .setPartitionCount(partitionCount)
+            .setChunkingEnabled(false)
+            .build());
+  }
+
+  // Visible for testing
+  protected VeniceWriter<byte[], byte[], byte[]> constructVeniceWriter(
+      Properties properties,
+      VeniceWriterOptions writerOptions) {
+    return new VeniceWriterFactory(properties).createVeniceWriter(writerOptions);
+  }
+
+  protected RecordSerializer<Object> getSerializer(Schema schema) {
+    return FastSerializerDeserializerFactory.getAvroGenericSerializer(schema);
+  }
+
+  private static Schema getSchemaFromObject(Object object) {
+    if (object instanceof GenericContainer) {
+      GenericContainer obj = (GenericContainer) object;
+      return obj.getSchema();
+    } else if (object instanceof CharSequence) { // convenience option.
+      return STRING_SCHEMA;
+    } else if (object instanceof Integer) {
+      return INT_SCHEMA;
+    } else if (object instanceof Long) {
+      return LONG_SCHEMA;
+    } else if (object instanceof Double) {
+      return DOUBLE_SCHEMA;
+    } else if (object instanceof Float) {
+      return FLOAT_SCHEMA;
+    } else if (object instanceof byte[] || object instanceof ByteBuffer) {
+      return BYTES_SCHEMA;
+    } else if (object instanceof Boolean) {
+      return BOOL_SCHEMA;
+    } else {
+      throw new VeniceException(
+          "Venice Producer only supports Avro objects, and primitives, found object of class: "
+              + object.getClass().toString());
+    }
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
+    String error = validateProducer();
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
+    String error = validateProducerWithLogicalTime(logicalTime);
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(logicalTime, key, value);
+  }
+
+  private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
+    producerMetrics.recordPutRequest();
+    return CompletableFuture.supplyAsync(() -> {
+      Schema valueSchema;
+      try {
+        valueSchema = getSchemaFromObject(value);
+      } catch (Exception e) {
+        producerMetrics.recordFailedRequest();
+        throw e;
+      }
+      // Might block
+      int valueSchemaId;
+      try {
+        valueSchemaId = schemaReader.getValueSchemaId(valueSchema);
+      } catch (Exception e) {
+        valueSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
```

Comment:
Are we just going to swallow this exception? I understand that we want to keep the amount of stuff we log minimal as a client library but perhaps we can afford to put the underlying exception name in the VeniceException message that we throw later in case the error persists and we need it for debugging.

Reply:
Makes sense. Will add this information